scripts/spring19_seminar/cola_npi_x.sh

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+#SBATCH --time=12:00:00
+#SBATCH --mem=20000
+#SBATCH --gres=gpu:1
+#SBATCH --job-name=myrun
+#SBATCH --output=slurm_%j.out
+
+# Naming Scheme
+# the experiment name is named after "npi_[MODEL NAME]", e.g., npi_bert
+# the run name falls into three kinds
+# (1) if the run goes through pretraining (e.g., ccg, mnli, or none) and some finetuning tasks and finally the target tasks,
+#     then the run name is "run_[MODEL NAME][PRETRAIN]_[FINETUNE NAME]";
+# (2) if the run merely does pretraining to generate a best model for loading in later runs,
+#     then the run name is "run_[MODEL NAME][PRETRAIN]_model";
+# (3) if the run loads a pretrained model and goes through finetuning tasks and then target tasks,
+#     then the run name is "run_[MODEL_NAME][PRETRAIN]_[FINTUNE NAME]", same as in (1).
+
+
+# Finetune-Target Tasks Combinations
+# All the finetune and target tasks are CoLA-like.
+# Finetuning on CoLA, we target CoLA and all nine NPI tasks listed below:
+
+#      - 'cola_npi_sup'            NPI licensed by superlatives
+#      - 'cola_npi_quessmp'        NPI licensed by simple questions  
+#      - 'cola_npi_ques'           NPI licensed by questions
+#      - 'cola_npi_qnt'            NPI licensed by quantifiers
+#      - 'cola_npi_only'           NPI licensed by "only"
+#      - 'cola_npi_negsent'        NPI licensed by sentential negation
+#      - 'cola_npi_negdet'         NPI licensed by determiner negation
+#      - 'cola_npi_cond'           NPI licensed by conditionals
+#      - 'cola_npi_adv'            NPI licensed by adverbials
+
+# These 9 different environments are to be found at https://github.com/alexwarstadt/data_generation/tree/master/outputs/npi/environments/splits
+# Finetuning on one of the NPI task, we target all NPI tasks;
+# Finetuning on eight of the NPI tasks, we target the remaining one.
+
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola, target_tasks = \"cola,cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola\", do_pretrain = 1" 
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_adv\", do_pretrain = 1" 
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_cond\", do_pretrain = 1" 
```

Comment:
We should probably separate pretraining and target-task training into two steps, since the current setting would make all the target_tasks fine-tune BERT, if I'm not mistaken.

Separation may be done like this:
python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"\", pretrain_tasks = \"cola_npi_adv\", do_pretrain = 1" 

python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", transfer_paradigm = \"finetune\""

Reply:
Currently all "fine-tuning tasks" have been treated as "pretraining tasks".


```python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_cond\", do_pretrain = 1" ```

The intention of the above cmd is like

(1) train (plain) BERT on cola_npi_cond;
(2) test BERT of (1) on cola_npi_sup;
(3) test BERT of (1) on cola_npi_quessmp;
...
(n) test BERT of (1) on cola_npi_adv.

Since do_target_task_training has been turned off, it might not be a problem.
I remembered @HaokunLiu  mentioned in Slack that even if we do_target_task_training, the execution is still in a "distributive" manner: for each cola_npi_x  BERT of (1) is finetuned on cola_npi_x and eval on cola_npi_x

Can @HaokunLiu  @sleepinyourhat  confirm this?

Reply:
@sheng-fu :  maybe I misunderstood something. But if do_target_task_training first fine-tunes the model on *all* target tasks before evaluate one after another, the following separation might still have the same problem

> `Separation may be done like this:
python main.py --config_file config/spring19_seminar/bert.conf 
--overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = "", pretrain_tasks = "cola_npi_adv", do_pretrain = 1"

python main.py --config_file config/spring19_seminar/bert.conf 
--overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = "cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv", transfer_paradigm = "frozen""`

Reply:
Oh I see, so your commands are just pre-training commands. The target-tasks do not really matter then? So I guess my point would be, in evaluation, we should make sure that transfer_paradigm is frozen.

    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", do_target_task_training = 1,  transfer_paradigm = \"frozen\""

----

By the way, all this relies on cgg/mnli and cola/npi tasks are both pretraining tasks and can maybe be pretrained one after another, right?

Reply:
Oh sorry I missed the part about doing full eval without doing target_task_training. Then you're right, what I raised is not a problem.

This still relies on bert+ccg and bert+mnli being able to undergo further pretraining, right? I tried today and couldn't get that to work. Should we just throw ccg and mnli as part of the pretraining tasks together with the cola/npi tasks?